azure/Guidelines.md

```diff
@@ -0,0 +1,865 @@
+> 
+> :white_check_mark: **DO** carefully craft ```x-ms-error-code``` string values for errors that are recoverable at runtime.
+> 
+> :white_check_mark: **DO** ensure that the top-level ```code``` field's value is identical to the ```x-ms-error-code``` header's value (see example body below).
+> 
+> :white_check_mark: **DO** document the service's error code strings; they are part of the API contract.
+> 
+> :white_check_mark: **DO** provide a response body as follows (example):
+```json
+{
+  "error": {
+    "code": "InvalidPasswordFormat",
+    "message": "Human-readable description",
+    "target": "target of error",
+    "innererror": {
+      "code": "PasswordTooShort",
+      "minLength": 6,
+    }
+  }
+} 
+```
+
+> :heavy_check_mark: **YOU MAY** group common customer code errors into a few ```x-ms-error-code``` string values.
+>
+> :heavy_check_mark: **YOU MAY** treat the other fields as you wish as they are __not__ considered part of your service's API contract and customers should not take a dependency on them or their value. They exist to help customers self-diagnose issues.
+
+
+### JSON
+Services, and the clients that access them, may be written in multiple languages. To ensure interoperability, JSON establishes the "lowest common denominator" type system, which is always sent over the wire as UTF-8 bytes. This system is very simple and consists of three types:
+
+ Type | Description
+ ---- | -----------
+ Boolean | true/false (always lowercase)
+ Number  | Signed floating point (IEEE-754 binary64; int range: -2<sup>53</sup>+1 to +2<sup>53</sup>-1)
+ String  | Used for everything else
+
+> :white_check_mark: **DO** use integers within the acceptable range of JSON number.
+> 
+> :white_check_mark: **DO** establish a well-defined contract for the format of strings. For example, determine maximum length, legal characters, case-(in)sensitive comparisons, etc. Where possible, use standard formats, e.g. RFC3339 for date/time.
+> 
+> :white_check_mark: **DO** use strings formats that are well-known and easily parsable/formattable by many programming languages, e.g. RFC3339 for date/time.
+> 
+> :white_check_mark: **DO** ensure that information exchanged between your service and any client is "round-trippable" across multiple programming languages.
+> 
+> :white_check_mark: **DO** use [RFC3339] for date/time.
+> 
+> :white_check_mark: **DO** use [RFC4122] for UUIDs.
+
+> :heavy_check_mark: **YOU MAY** use JSON objects to group sub-fields together.
+> 
+> :heavy_check_mark: **YOU MAY** use JSON arrays if maintaining an order of values is required.
+
+> :ballot_box_with_check: **YOU SHOULD** use JSON objects instead of arrays whenever possible.
+
+> :warning: **YOU SHOULD NOT** use JSON Arrays, e.g. [ value, … ]. Arrays are very difficult and inefficient to work with, especially with ```JSON Merge Patch```, as the entire array needs to be read prior to any operation being applied to it.
```

Comment:
I would move this closer to the MAY use JSON arrays and clarify the conflict between MAY and SHOULD NOT

Reply:
Got it--I can see the confusion here. And will clarify.